app/assets/src/components/views/PipelineViz.jsx

```diff
@@ -0,0 +1,337 @@
+    const zoomChange = (e.deltaY < 0 ? 1 : -1) * this.props.zoomChangeInterval;
+    this.setState({ zoom: this.state.zoom + zoomChange });
+  }
+
+  toggleStage(index) {
+    const updatedStagesOpened = [...this.state.stagesOpened];
+    updatedStagesOpened[index] = !updatedStagesOpened[index];
+    this.setState({ stagesOpened: updatedStagesOpened });
+  }
+
+  modifyStepNames() {
+    // Strips 'PipelineStep[Run/Generate]' from front of each step name.
+    // TODO(ezhong): Consider adding 'name' field to dag_json later.
+    forEach(this.props.stageResults.stages, (stageData, _) => {
+      stageData.steps.forEach(step => {
+        if (step.class.substring(0, 12) == "PipelineStep") {
+          step.class = step.class.substring(12);
+        }
+        if (step.class.substring(0, 3) == "Run") {
+          step.class = step.class.substring(3);
+        }
+        if (step.class.substring(0, 8) == "Generate") {
+          step.class = step.class.substring(8);
+        }
+      });
+    });
+  }
+
+  populateNodeAndEdgeData(index, nodeData, edgeData) {
+    const stageData = this.props.stageResults.stages[this.stageNames[index]];
+    const stepData = stageData.steps;
+
+    const outTargetOrPathToStepId = {};
+    const currFileNameToOutputtingNode = {};
+    stepData.forEach((step, i) => {
+      // Populate nodeData
+      nodeData.push({ id: i, label: step.class, level: 1 });
+
+      if (!(step.out in outTargetOrPathToStepId)) {
+        // Populate outFileToStepId for intra-stage edges
+        outTargetOrPathToStepId[step.out] = i;
+        // Populate currFileNameToOutputtingNode for inter-stage edges
+        stageData.targets[step.out].forEach(fileName => {
+          const fileNameWithPath = `${stageData.output_dir_s3}/${
+            this.pipelineVersion
+          }/${fileName}`;
+          currFileNameToOutputtingNode[fileNameWithPath] = i;
+        });
+      }
+    });
+
+    // Add beginning (input) and ending (output) nodes
+    nodeData.push({ id: START_NODE_ID, level: 0, group: "startEndNodes" });
+    nodeData.push({ id: END_NODE_ID, group: "startEndNodes" });
+
+    let maxLevel = 1;
+    stepData.forEach((step, i) => {
+      step.in.forEach(inTarget => {
+        if (inTarget in outTargetOrPathToStepId) {
+          const fromId = outTargetOrPathToStepId[inTarget];
+          edgeData.push({ from: fromId, to: i });
+
+          nodeData[i].level = Math.max(
+            nodeData[i].level,
+            nodeData[fromId].level + 1
+          );
+          maxLevel = Math.max(maxLevel, nodeData[i].level);
+        } else {
+          // Connect beginning steps to input node
+          edgeData.push({ from: START_NODE_ID, to: i });
+        }
+      });
+    });
+    nodeData[stepData.length + 1].level = maxLevel + 1;
+
+    if (index < this.stageNames.length - 1) {
+      // Create edges to output node if it's output files appear in next stage's inputs.
+      const nextStageData = this.props.stageResults.stages[
```

Comment:
not a blocker but we still want to pull out all the props at the top of the function like `const { stageResults, backgroundColor } = this.props`